src/ticks/debugger_mi2.c

```diff
@@ -0,0 +1,1359 @@
+            // ignore next option
+            i++;
+        } else {
+            if (expression == NULL) {
+                expression = strdup(arg);
+            } else {
+                char spaced[256];
+                sprintf(spaced, "%s %s", expression, arg);
+                free(expression);
+                expression = strdup(spaced);
+            }
+        }
+    }
+
+    if (expression == NULL) {
+        mi2_printf_error("expression is not specified");
+        return;
+    }
+
+    bk.debug("evaluating expression %s\n", expression);
+    evaluate_expression_string(expression);
+
+    struct expression_result_t* result = get_expression_result();
+    if (is_expression_result_error(result))
+    {
+        mi2_printf_error("%s", result->as_error);
+        free(expression);
+        return;
+    }
+
+    char evaluated_value[128] = "<unknown>";
+    expression_result_value_to_string(result, evaluated_value, sizeof(evaluated_value));
+
+    mi2_printf_response("done,value=\"%s\"", evaluated_value);
+    free(expression);
+}
+
+static char* sprintf_frame0(char* ptr) {
+    struct debugger_regs_t regs;
+    bk.get_regs(&regs);
+
+    debug_frame_pointer* fp = debug_stack_frames_construct(regs.pc, regs.sp, &regs, 1);
+
+    if (fp) {
+        if (fp->symbol && fp->filename && fp->function) {
+            ptr += sprintf(ptr,
+                "level=\"0\",addr=\"0x%08x\",func=\"%s\","
+                "args=[],file=\"%s\","
+                "fullname=\"%s\",line=\"%d\"",
+                regs.pc, fp->function->name, fp->filename, fp->filename, fp->lineno);
+            goto done;
+        }
+    } else {
+        uint16_t offset;
+        symbol* sym = symbol_find_lower(regs.pc, SYM_ADDRESS, &offset);
+        if (sym != NULL) {
+            const char *filename;
+            int lineno;
+            if (debug_find_source_location(regs.pc, &filename, &lineno)) {
+                ptr += sprintf(ptr, "level=\"0\",addr=\"0x%08x\",func=\"%s\","
+                                    "file=\"%s\","
+                                    "fullname=\"%s\",line=\"%d\","
+                                    "arch=\"z80\"", regs.pc, sym->name, sym->file, filename, lineno);
+            } else {
+                ptr += sprintf(ptr, "level=\"0\",addr=\"0x%08x\",func=\"%s\","
+                                    "file=\"%s\","
+                                    "fullname=\"%s\","
+                                    "arch=\"z80\"", regs.pc, sym->name, sym->file, sym->file);
+            }
+        } else {
+            ptr += sprintf(ptr, "level=\"0\",addr=\"0x%08x\"", regs.pc);
+        }
+    }
+
+done:
+    debug_stack_frames_free(fp);
+    return ptr;
+}
+
+static void cmd_stack_list_frames(int argc, char **argv) {
+
+    struct debugger_regs_t regs;
+    bk.get_regs(&regs);
+
+    uint16_t stack = regs.sp;
+    uint16_t initial_stack = stack;
+    uint16_t at = bk.pc();
+
+    char* dump_buffer = malloc(65536);
```

Comment:
Using utstring removes the potential/theoretical buffer overflow

Reply:
Thanks, incredible tool!